diffusion-model-with-mnist-data.py

```diff
@@ -167,3 +167,53 @@ def forward(self, noisy_images, timestep):
 # # Evaluation
 
 
+# -
+
+# # Sampling
+
+# +
+def sample_image(image_size, beta, noise_predictor, simple_variance=False):
+    "sample image"
+    
+    timesteps = beta.shape[0] - 1
+    alpha = 1 - beta
+    alpha_cum = torch.cumprod(alpha, dim=0)
+    if simple_variance:
+        variances = beta
+    else:
+        alpha_cum_t = torch.cat(alpha_cum, torch.Tensor([0]))
+        alpha_cum_t_minus_1 = torch.cat(torch.Tensor([0]), alpha_cum)
+        variances = (1-alpha_cum_t_minus_1)/(1-alpha_cum_t)
+        variances = variances[:-1] * beta
+    
+    x_t = torch.normal(torch.zeros(image_size), torch.ones(image_size))
+    
+    for timestep in range(timesteps, -1, -1):
+        predicted_noise = noise_predictor(x_t, timestep)
+        z = torch.normal(torch.zeros(image_size), torch.ones(image_size))
+        
+        x_t = variances[timestep] * z + (x_t - (1-alpha[timestep])/torch.sqrt(1-alpha_cum[timestep])*predicted_noise)
```

Comment:
The coefficient 1 / torch.sqrt(alpha[timestep]) before (x_t - ...) is missing. Add this.